OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyUtils.cs

```diff
@@ -73,6 +73,44 @@ public SlotDropDownOption(string title, string order, Func<bool> selected)
 			dropdown.ShowDropDown("LABEL_DROPDOWN_TEMPLATE", 167, options, setupItem);
 		}
 
+		public static void ShowPlayerActionDropDown(DropDownButtonWidget dropdown, Session.Slot slot,
+			Session.Client c, OrderManager orderManager, Widget lobby,  Action before, Action after)
+		{
+			Action<bool> okPressed = tempBan => { orderManager.IssueOrder(Order.Command("kick {0} {1}".F(c.Index, tempBan))); after(); };
+			var onClick = new Action(() =>
+			{
+				before();
+
+				Game.LoadWidget(null, "KICK_CLIENT_DIALOG", lobby.Get("TOP_PANELS_ROOT"), new WidgetArgs
+				{
+					{ "clientName", c.Name },
+					{ "okPressed", okPressed },
+					{ "cancelPressed", after }
+				});
+			});
+
+			var options = new List<ShowPlayerActionDropDownOption>
+			{
+				new ShowPlayerActionDropDownOption("Promote to Admin", () => { orderManager.IssueOrder(Order.Command("promote {0}".F(c.Index))); }),
+				new ShowPlayerActionDropDownOption("Kick", onClick),
+				new ShowPlayerActionDropDownOption("Move to spectator", () => { orderManager.IssueOrder(Order.Command("move_to_spectators {0}".F(c.Index))); })
+				{
+					IsVisible = () => !c.IsObserver && orderManager.LobbyInfo.GlobalSettings.AllowSpectators
```

Comment:
This leaves a hole when opening the menu on a spectator:
![screen shot 2018-06-30 at 17 00 56](https://user-images.githubusercontent.com/167819/42126916-75933574-7c87-11e8-9a4b-73839ec73288.png)

Define the list with just the Kick option, then add this and the admin option to `options` only if the appropriate conditions are true.